README.md
thp
===

Tiny Honeypot

Created a Github Repo for archival purposes for tiny honeypot.

I accessed the files from the wayback machine. You can take a look at the page here: http://web.archive.org/web/20090827091208/http://www.alpinista.org/thp/

Be warned this a really old project but I felt this would be a good project to resurrect.

Quote/Description from page:
Tiny Honeypot - resource consumption for the good guys

They've been filling our logs, wires, and storage for years; now it's time to turn the tables around.
Wouldn't it be nice if every single unsolicited connection attempt tied up the attacker who launched it by appearing to actually work, all the while providing a little insight into their motives & intents? thp appears to listen on all ports otherwise not in legitimate use, providing a series of phony responses to attacker commands. Some are very simple, others are somewhat more interactive. The goal isn't to fool a skilled, determined attacker...merely to cloud the playing field with tens of thousands of fake services, all without causing unreasonable stress on the thp host.

As an addition to your state and content-aware Intrusion Detection System, ie: Snort, thp allows nearly every connection attempt to complete, thus your content rules have a chance to actually fire, rather than depending on simple port and protocol "context" filters.

The wonderfully robust & flexible Linux netfilter is used with the REDIRECT target used to send any incoming connection requests to the thp listener(s), which are managed by xinetd. This does limit the hosts that can play to Linux v2.4.x systems, but stateful firewalling that can do "many -> one" destination network address translation is essential. This concept could be extended to other platforms/firewalls with similar capabilities; I know of one such OpenBSD/ipf box...Dan, can I include your scripts?

IMPORTANT NOTE: You can certainly run this on hosts that are in daily use with a negligible impact on performance, but please be sure to customize the firewall script to suit your environment. If you aren't careful, any existing firewall configuration could be replaced by thp's default rules.
The latest version is 0.4.6.


CHANGELOG
0.4.6 - Enhanced shell.pl to include a working directory array, makes pwd 
and cd work. Added wget, more flexible uname (with -a, -s, -m, -v), w, and 
support for semicolon separated command list.

0.4.5 (not released) - Added Time::HiRes as the best method of pulling good 
timestamps. Moved thpfunc.pl into ~thp/lib. Improvements to iptables script, 
primarily to fix a firewall logging bug.
Chris Corella : Added (thp.conf)$uname mostly to simulate how easy it is to 
add shell like capabilities. Added (shell.pl) select LOG;$|=1; to set LOG to
currently selected output filehandler, and flush buffer on all PRINT 
statements to LOG. Added (shell.pl) uname -a, nad whoami functionality, see 
first comment. Added (all handlers) select LOG; $|=1; code

0.4.4 - Lots of changes here! Capture logs now include src address & port
of attacker. Selectable in the config file, capture log entries on a 
single-line, syslog-style, suitable for machine parsing. Check the comments
for certain implications of choosing this option.
I changed the directory layout a bit with this release. In the $thpdir/lib
directory are the individual functions libs, as well as other response
files.  Http.pl is completely rewritten, with the goal of rfc2616 compliance 
whenever possible. It includes errors 414, 501, 400 & 414, correctly built
http return headers for several file types, chameleon mode - to change its
responses (if turned on) to respond like an IIS machine when an attacker
requests certain types of resources, and several other features. It is 
useable (if you are a little deranged) as a lightweight webserver with a
flat directory structure.

0.4.3-2 - adjusted xinetd.d file port numbers & removed o-x on conf files
(pr0ps to Bill Scherr), added GOODNET & GOODSVCS to INPUT chain, added a 
section to iptables.rules to allow a multi-homed system to trust either an
entire interface, or just a trusted network. Added a test to bomb out if
someone accidentaly runs iptables.rules as-is on a machine that is a router.
Fixed escapes & array references in ftp() that were causing some versions 
of Perl to complain.

0.4.3-1 - Fixed extra shell prompt on exit, added GPL blurb to all files 
(thanks to Scot Wilcoxon for pointing this out), removed duplicate xinetd.d
 files from the tarball (sorry folks), iptables script requires less post-
install tweaking for hpot_svcs, moved port range for listeners to 40k+ to 
avoid conflicts with fakerpc, several other little tweaks & bugfixes.

0.4.3 - Added session timeouts, simple http emul, pid on caplog start line
 (you can now easily correlate with xinetd logging), xinetd per-source lim
its by default.

0.4.2-2 - Tweaked ftp some more, added a thp.conf switch to allow genuine
data connections for ftp, thus STOR functions and deposits the file(s) 
uploaded into the hpot directory. LIST & RETR just get junk via greetbin. 
Provided for a silent listener (no prompt or greeting). Various 
housekeeping. 

0.4.2-1 - Tweaked ftp a bit, pasv, port, list, retr now do something.

0.4.2 - Added an ftp responder in thpfunc.pl. Improved session naming by 
using the syscall gettimeofday and concatenating seconds & usecs in hex, 
with a fallback to the old method. Added a nullresponder sub to 
thpfunc.pl.

0.4.1 - Due to a piss poor job of input validation (none), thus the ease 
with which an intruder could execute arbitrary commands as user "nobody", 
I have begun a rewrite in Perl, taking a little better care in examining (
or not evaluating at all) any external data exchanges. Broke things up 
into thp.cfg (config stuff), thpfunc.pl (functions lib), and relegated 
logthis to dealing with start/stop logging & calling the service-specific 
subroutines (not there yet).

0.3.2 - Extended iptables script to simplify multiple hpot_svcs.  
Clarified and enhanced iptables support for permitted (trusted) networks 
and/or services. Started playing with thpsvcs, to provide believeable 
responses to simple initial commands. ftp is the only one in there now, 
more to come.

0.3.1 - Some xinetds don't like the config file name xinetd.hpot, so I 
renamed it to simply hpot. Added the necessary iptables lines to allow 
local -> external & return trip traffic.  Fixed README (yeah, right).

0.3 - First real build.  Prior versions were local cob-jobs. I figure 0.3 
is about right.


lib/ftpd.pl
sub ftp {
srand(time);
my $ftpuser = "anon";
my $type = I;
my @addroct = split /\./, $thpaddr;
#my @portoct = ((($shorttime % 124) + 4),($shorttime % 255));
unless (defined $pasvport){
	$pasvport = int(rand 65535) + 1025;
}
my @pasvoct = ($pasvport >> 8, $pasvport & 0xff);
my $file = "file";
%ftphash = (
	user	=>	"331 Password required for $ftpuser\x0d\x0a",
	pass	=>	"230 User $ftpuser logged in.\x0d\x0a",
	already	=>	"530 Already logged in.\x0d\x0a",
	nologin	=>	"530 Please login with USER and PASS.\x0d\x0a",
	start	=>	"220 $hostname.$domain $ftpver ready.\x0d\x0a",
	syst	=>	"215 UNIX Type: L8\x0d\x0a",
	pwd	=>	"257 \"/\" is current directory.\x0d\x0a",
	type	=>	"200 Type set to $type.\x0d\x0a",
	mkd	=>	"257 New directory created.\x0d\x0a",
	stor	=>	"150 Opening BINARY mode data connection.\x0d\x0a",
	pwd	=>	"257 \"/\" is current directory.\x0d\x0a",
	cwd	=>	"250 CWD command successful.\x0d\x0a",
	cdup	=>	"257 \"/\" is current directory.\x0d\x0a",
	port	=>	"500 Passive mode only.\x0d\x0a",
	port502	=>	"502 Illegal PORT Command\x0d\x0a",
	port200	=>	"200 PORT command successful.\x0d\x0a",
	actv425	=>	"425 Can't build data connection: Connection refused.\x0d\x0a",
	compl	=>	"226 Transfer complete.\x0d\x0a",
	rnfr	=>	"350 File exists, ready for destination name.\x0d\x0a",
	rnto	=>	"250 RNTO command successful.\x0d\x0a",
	retr	=>	qq (150 Opening ASCII mode data connection for \'$file\'.\x0d\x0a),
	list	=>	qq (150 Opening ASCII mode data connection for 'file list'.\x0d\x0a),
	pasv	=>	qq (227 Entering Passive Mode \($addroct[0],$addroct[1],$addroct[2],$addroct[3],$pasvoct[0],$pasvoct[1]\)\x0d\x0a),
	help	=>	qq (214-The following commands are recognized.
   USER    PORT    STOR    RNTO    NLST    MKD     CDUP 
   PASS    PASV    APPE    ABOR    SITE    XMKD    XCUP 
   TYPE    DELE    SYST    RMD     STOU 
   STRU    ALLO    CWD     STAT    XRMD    SIZE 
   MODE    REST    XCWD    HELP    PWD     MDTM 
   QUIT    RETR    RNFR    LIST    NOOP    XPWD 
214 Direct comments to root\@localhost.\x0d\x0a),

	"site help" =>	qq (214-The following SITE commands are recognized.
   UMASK   CHMOD   GROUP   NEWER   INDEX   ALIAS   GROUPS 
   IDLE    HELP    GPASS   MINFO   EXEC    CDPATH 
214 Direct comments to root\@localhost.\x0d\x0a),

	quit	=>	qq (221-You have transferred 0 bytes in 0 files.
221-Total traffic for this session was 2164 bytes in 0 transfers.
221 Thank you for using the FTP service on $hostname.$domain.\x0d\x0a)
);
	
  $login = 0;
  print STDERR $ftphash{start};
  while (my $commands = <STDIN>) {
    open(LOG, ">>$sesslog");
    select LOG;
    $|=1;
    print LOG $commands;
    chomp $commands;
    $commands =~ s/\r//;
    @commands=split /\s+/,($commands);

    if ($commands[0] =~ /user/i && $commands[1] =~ /[[:alnum:]]+/){
	if ($login == 1) {
	  print STDERR $ftphash{already};
	} else {
	  $ftpuser = $commands[1];
	  $ftphash{user} =~ s/anon/$ftpuser/;
	  $ftphash{pass} =~ s/anon/$ftpuser/;
	  print STDERR $ftphash{user};
	}

    } elsif ($commands[0] =~ /pass/i && $commands[1] =~ /[[:print:]]+/) {
	if ($login == 1) {
          print STDERR $ftphash{already};
        } else { 
	  if ($ftpuser) {
	    $login = 1;
	    print STDERR $ftphash{pass};
	  }
	}

    } elsif ($commands[0] =~ /list|retr|stor/i) {
        if ($login == 1) {
	  $commands[0] =~ tr/A-Z/a-z/;
		if (defined ($actvport)) {
		  $retval = active($commands[0], $commands[1]);
		  print STDERR $ftphash{$retval};
		} else {
	          print STDERR $ftphash{$commands[0]};
        	  sleep 1;
	          print STDERR $ftphash{compl};
		}
	} else {
	  print STDERR $ftphash{nologin};
        }

   } elsif ($commands[0] =~ /help|pasv|pwd|syst|rnfr|rnto|mkd|cwd|cdup|type/i) {
        if ($login == 1) {
	  $commands[0] =~ tr/A-Z/a-z/;
          print STDERR $ftphash{$commands[0]};
	} else {
	  print STDERR $ftphash{nologin};
        }
    } elsif ($commands[0] =~ /port/i) {
        if ($login == 1) {
          $success = ftpport($commands[1]) 
		if ($commands[1] =~ /(\d){1,3},(\d){1,3},(\d){1,3},(\d){1,3},(\d){1,3},(\d){1,3}/);
	  print STDERR $ftphash{$success};
	  $actvport = 1;
	} else {
	  print STDERR $ftphash{nologin};
        }

    } elsif ("$commands" =~ /\bsite help\b/i) {
        if ($login == 1) {
	  $commands =~ tr/A-Z/a-z/;
          print STDERR $ftphash{"$commands"};
	} else {
	  print STDERR $ftphash{nologin};
        }

   } elsif ($commands[0] =~ /exit\b|quit\b/i) {
	print STDERR $ftphash{quit};
        return;

    } else {
	if ($login == 1) {
	  print STDERR "500 @commands: command not understood.\x0d\x0a";
	} else {
	print STDERR $ftphash{nologin};
	}
    }
    close LOG;
  }
}



lib/smtp.pl
sub smtp {

$now = strftime("%a, %B %d %Y %T GMT", gmtime(time));

open (HTAB, "$thpdir/lib/smtptab");
my @keys = qw(
	state
	Command
	regex	
	newstate
	continue
	response
	assignment
);
$cnt = 0;

while (<HTAB>) {
 unless ( /^$|^#/ ) {
	chomp;
	my $key;
 	$cnt++;
	$strcnt = sprintf (qq(%0.2d), $cnt);
	@_ = split(/\t/, $_, 7);

 	foreach $key (@keys){
		$rules{"$strcnt$key"} = shift @_;
	}
 }
}

foreach $k (sort keys %rules) {
    print "$k => $rules{$k}\n";
}

close HTAB;

%smtp = (
	start	=>	"220 $hostname.$domain ESMTP Sendmail 8.11.2/8.11.2; $now\x0d\x0a",
	helo	=>	"250 $hostname.$domain Hello $dom [$saddr], pleased to meet you\x0d\x0a",
	err501	=>	"501 5.0.0 Invalid domain name\x0d\x0a",
	ehlo	=>	qq (250 $hostname.$domain Hello $dom [$saddr], pleased to meet you
250-ENHANCEDSTATUSCODES
250-8BITMIME
250-SIZE
250-DSN
250-ONEX
250-ETRN
250-XUSR
250-AUTH GSSAPI
250 HELP
),
	err503	=>	"503 5.0.0 $hostname.$domain Duplicate HELO/EHLO\x0d\x0a",
	mail	=>	"250 2.1.0 $rpath... Sender ok\x0d\x0a",
	already	=>	"503 5.5.0 Sender already specified\x0d\x0a",
	bogs	=>	"500 5.5.1 Command unrecognized: \"$cmd\"\x0d\x0a",
	err553	=>	"553 5.1.0 ... prescan: token too long\x0d\x0a",
	norp	=>	"503 5.0.0 Need MAIL command\x0d\x0a",
	nofp	=>	"503 5.0.0 Need RCPT (recipient)\x0d\x0a",
	lrcpt	=>	"250 2.1.5 $lrcpt... Recipient ok\x0d\x0a",
	rrcpt	=>	"250 2.1.5 $rrcpt... Recipient ok (will queue)\x0d\x0a",
	data	=>	"354 Enter mail, end with \".\" on a line by itself\x0d\x0a",
	eof	=>	"250 2.0.0 $qid Message accepted for delivery\x0d\x0a",
	help	=>	qq (214-2.0.0 This is sendmail version 8.11.2
214-2.0.0 Topics:
214-2.0.0       HELO    EHLO    MAIL    RCPT    DATA
214-2.0.0       RSET    NOOP    QUIT    HELP    VRFY
214-2.0.0       EXPN    VERB    ETRN    DSN     AUTH
214-2.0.0       STARTTLS
214-2.0.0 For more info use "HELP <topic>".
214-2.0.0 To report bugs in the implementation send email to
214-2.0.0       sendmail-bugs@sendmail.org.
214-2.0.0 For local information send email to Postmaster at your site.
214 2.0.0 End of HELP info\x0d\x0a),
	ehlohlp	=>	qq (214-2.0.0 EHLO <hostname>
214-2.0.0       Introduce yourself, and request extended SMTP mode.
214-2.0.0 Possible replies include:
214-2.0.0       SEND            Send as mail                    [RFC821]
214-2.0.0       SOML            Send as mail or terminal        [RFC821]
214-2.0.0       SAML            Send as mail and terminal       [RFC821]
214-2.0.0       EXPN            Expand the mailing list         [RFC821]
214-2.0.0       HELP            Supply helpful information      [RFC821]
214-2.0.0       TURN            Turn the operation around       [RFC821]
214-2.0.0       8BITMIME        Use 8-bit data                  [RFC1652]
214-2.0.0       SIZE            Message size declaration        [RFC1870]
214-2.0.0       VERB            Verbose                         [Allman]
214-2.0.0       ONEX            One message transaction only    [Allman]
214-2.0.0       CHUNKING        Chunking                        [RFC1830]
214-2.0.0       BINARYMIME      Binary MIME                     [RFC1830]
214-2.0.0       PIPELINING      Command Pipelining              [RFC1854]
214-2.0.0       DSN             Delivery Status Notification    [RFC1891]
214-2.0.0       ETRN            Remote Message Queue Starting   [RFC1985]
214-2.0.0       STARTTLS        Secure SMTP                     [RFC2487]
214-2.0.0       AUTH            Authentication                  [RFC2554]
214-2.0.0       XUSR            Initial (user) submission       [Allman]
214-2.0.0       ENHANCEDSTATUSCODES     Enhanced status codes   [RFC2034]
214 2.0.0 End of HELP info\x0d\x0a),
	quit	=>	qq (221 2.0.0 $hostname.$domain closing connection\x0d\x0a)
);
	
  $login = 0;
  print STDERR $smtphash{start};
  while (my $commands = <STDIN>) {
    open(LOG, ">>$sesslog");
    print LOG $commands;
    select LOG;
    $|=1;
    chomp $commands;
    $commands =~ s/\r//;
    @commands=split /\s+/,($commands);

    if ($commands[0] =~ /user/i && $commands[1] =~ /[[:alnum:]]+/){
	if ($login == 1) {
	  print STDERR $ftphash{already};
	} else {
	  $ftpuser = $commands[1];
	  $ftphash{user} =~ s/anon/$ftpuser/;
	  $ftphash{pass} =~ s/anon/$ftpuser/;
	  print STDERR $ftphash{user};
	}

    } elsif ($commands[0] =~ /pass/i && $commands[1] =~ /[[:print:]]+/) {
	if ($login == 1) {
          print STDERR $ftphash{already};
        } else { 
	  if ($ftpuser) {
	    $login = 1;
	    print STDERR $ftphash{pass};
	  }
	}

    } elsif ($commands[0] =~ /list|retr|stor/i) {
        if ($login == 1) {
	  $commands[0] =~ tr/A-Z/a-z/;
          print STDERR $ftphash{$commands[0]};
	  sleep 1;
	  print STDERR $ftphash{compl};
	} else {
	  print STDERR $ftphash{nologin};
        }

   } elsif ($commands[0] =~ /help|pasv|port|pwd|syst|rnfr|rnto|mkd|cwd|cdup|type/i) {
        if ($login == 1) {
	  $commands[0] =~ tr/A-Z/a-z/;
          print STDERR $ftphash{$commands[0]};
	} else {
	  print STDERR $ftphash{nologin};
        }

    } elsif ("$commands" =~ /\bsite help\b/i) {
        if ($login == 1) {
	  $commands =~ tr/A-Z/a-z/;
          print STDERR $ftphash{"$commands"};
	} else {
	  print STDERR $ftphash{nologin};
        }

   } elsif ($commands[0] =~ /exit\b|quit\b/i) {
	print STDERR $ftphash{quit};
        return;

    } else {
	if ($login == 1) {
	  print STDERR "500 @commands: command not understood.\x0d\x0a";
	} else {
	print STDERR $ftphash{nologin};
	}
    }
    close LOG;
  }
}



lib/shell.pl
sub shell {
  $thpath = "$homedir";
  if ($thpath =~ m/^.*\/([^\/]+)/){
      $pathsuffix = $1;
  }
  if ($hostname =~ m/^([^\.]+)\./){
      $shortname = $1;
  }
  $prompt = "[root\@${shortname} ${pathsuffix}]# ";
  open(GREETING, "$greetbin|");
  while(<GREETING>) {
    print STDERR $_;
  }
  close(GREETING);

  print STDERR "$prompt";

  while (my $commandline = <STDIN>) {
  
    open(LOG, ">>$sesslog");
    select LOG;
    $|=1;
    print LOG $commandline;
    chomp $commandline;
    @commandline = (split ";", $commandline);

    while (@commandline){
        $commands = shift (@commandline) ;
    
        @command=split /\s+/,($commands);

        shift @command if ($command[0] =~ /^(\s|$)/);

        if ($command[0] =~ /\buname\b/){
            if ($command[1] =~ /-[amsv]\b/){
            print STDERR $shellhash{$command[0]}{$command[1]}, "\n";
            }else{
            print STDERR $shellhash{$command[0]}{-s}, "\n";
            }

        } elsif ($command[0] =~ /\bcd\b/){
            changedir("$command[1]");  

        } elsif ($command[0] =~ /\bpwd\b/){
            print STDERR "$thpath\n";

        } elsif ($command[0] =~ /\b(whoami|w|id|wget)\b/){
            print STDERR "$shellhash{$command[0]}\n";

        } elsif ($command[0] =~ /\b(exit|logout)\b/){
            close(LOG);
            return;

        }  
      }
	    print STDERR "$prompt";
    }
  }

  sub changedir{
      $elements = $_[0] or $elements = "$homedir";
      if ($elements =~ /^\/.*/){
          $elements =~ s/\///;
          @thpath = ();
      } else {
      @thpath = (split /\//, $thpath);
      }
      @elements = (split /\//, $elements);
      foreach $element (@elements){
          if ($element eq ".."){
             pop @thpath;
          } else {
             push @thpath, $element;
          }
      }
      $thpath = "/" . (join "/", @thpath);
      $thpath =~ s/^\/\//\//;
      $pathsuffix = pop @thpath;
      push @thpath, $pathsuffix;
      $pathsuffix = "/" unless $pathsuffix;
      $prompt = "[root\@$shortname $pathsuffix]# ";
  }
  sub hostname{
  }

%shellhash = (
	uname	=>	{
		-a	=>	"Linux localhost 2.2.17 #4 Mon Apr 7 09:04:33 EDT 2001 i686 unknown unknown GNU/Linux",
		-m 	=>	"i686",
		-s 	=>	"Linux",
		-v 	=>	"#4 Mon Apr 7 09:04:33 EDT 2001",
	},
	whoami	=>	"root",
	w	=>	  "3:32am  up  7:45, 10 users,  load average: 0.04, 0.05, 0.01
USER     TTY      FROM              LOGIN@   IDLE   JCPU   PCPU  WHAT",
	id	=>	"uid=0(root) gid=0(root) groups=0(root),1(bin),2(daemon),3(sys),4(adm),6(disk),10(wheel)",
	wget	=>	"wget: missing URL
Usage: wget [OPTION]... [URL]...

Try `wget --help' for more options.",
);



lib/ftpport.pl
sub ftpport {

use IO::Socket;
my $portspec = shift;
my @portspec = split (/\,/, $portspec);

$rhost = "$portspec[0].$portspec[1].$portspec[2].$portspec[3]";
$rport = (($portspec[4] << 8) + $portspec[5]);

if (inet_aton($rhost) ne inet_aton($saddr)) {
	return "port502";
} else {
	return "port200";
}
}

sub active {

use IO::Socket;

%actvhash = (
	dir150	=>	"150 Opening ASCII mode data connection for directory listing.\x0d\x0a",
	retr150	=>	"150 Opening BINARY mode data connection for $arg.\x0d\x0a"
	stor150	=>	"150 Opening BINARY mode data connection for $arg.\x0d\x0a"
);

my $actvcmd = shift;
my $arg = shift;

my $sock = IO::Socket::INET -> new(PeerAddr => "$rhost",
				PeerPort => "$rport",
				Proto	 => "tcp"
)
   or return "actv425";

if ($actvcmd =~ /stor/i) {
	print STDERR $actvhash{stor150};
} elsif ($actvcmd =~ /retr/i) {	
	print STDERR $actvhash{retr150};
}

my ($upload, $booty);
return unless (defined ($upload = fork()));
if ($upload) {
	while ($booty = <$sock>) {
		print LOG $booty;
		LOG->autoflush(1);
	}
	kill ('TERM', $upload);
} else {
	while (<STDIN>) {
		print $sock $_;
	}
}

close $sock;
return "compl";
}


lib/thpfunc.pl
# /usr/local/thp/thpfunc.pl version 0.4.4

# Functions for use in thp 0.4.x  A component of the thp
# honeypot kit.
#
# Copyright George Bakos - alpinista@bigfoot.com
# July 15, 2002
# This is free software, released under the terms of the GNU General 
# Public License avaiable at http://www.fsf.org/licenses/gpl.txt


sub getip {
$reply = `/sbin/ifconfig $intf`;
if ($reply =~ /^.*?\b(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})\b.*/is) {
  $thpaddr = $1
}
}

# Since our SIDs are hex concatanations of unix time in seconds & microseconds,
# we need a way to pull hi-resolution timestamps. Otherwise, we settle for 
# one-second accuracy, possibly leading to some mangled session logging.
# If Time::HiRes is available, our lives are easy. If not, lets see if the
# necessary headers are available for a gettimeofday() syscall. If that
# bombs too, we're stuck with plain ol' time. :-p

sub gettime {
if ( eval "require Time::HiRes" ) { 
  import Time::HiRes ;
  my ($secs, $usecs) = Time::HiRes::gettimeofday();
  $timestp = sprintf ("%.X%.X", ("$secs", "$usecs"));
  $shorttime = $secs;
} elsif (eval "require 'sys/syscall.ph'") {
  my $now = pack("LL", ());
  syscall( &SYS_gettimeofday, $now, undef) >= 0
        or die "gettimeofday: $!";
  my($secs, $usecs) = unpack("LL", $now);
  $timestp = sprintf ("%.X%.X", ("$secs", "$usecs"));
  $shorttime = $secs;
} else {
$shorttime = $timestp = time();
}
}

# signal handlers

# Use a SIGALRM to limit time of execution of each script
# Since $sid is only used to label the caplog entry (once
# things get going) we can here add a comment to it and exit
# with a nonzero value.
# It's a bit of a kludge; please improve on this, folks.

sub closeout {
  $sid = "$sid - timeout";
  clcaplog();
  close(CAPLOG);
  exit 5;
}
$SIG{ALRM} = \&closeout;

# Here, we manage the caplog file, which tracks all sessions

sub opncaplog {
  gettime();
  $start = $shorttime;
  $sid = $timestp;

  if ($svcname) {
        $sid="$sid.$svcname"}

  $sesslog="$logdir/$sid";

  if ($logtype eq "single") {
  @capdata = ((strftime("%b %d %T", localtime(time))), ("SID=$sid"), ("PID=$procid"), ("SRC=$saddr"), ("SPT=$sport"));
  } else { print (CAPLOG "\n", strftime("%b %d %T", localtime(time)), " start thp SID $sid, UNIX pid $procid source $nsdata[4]\n");
  }
}


sub clcaplog {
  gettime();
  $end = $shorttime;
  $eltime = $end - $start;

  if ($logtype ne "single") {
    print CAPLOG strftime("%b %d %T", localtime(time)), " end thp SID $sid\n";
  }

  if ($eltime > 0) {
    $etstr = (strftime("%T", gmtime($eltime)));
    push (@capdata,("ET=$etstr"));
    if ($logtype ne "single") {
      print CAPLOG "\t- elapsed time ", $etstr, "\n";
    }
  }

  if ($size=(-s $sesslog)) {
    push (@capdata,("BYTES=$size"));
    if ($logtype ne "single") {
      print CAPLOG "\t- total $size bytes\n";
    }
  }
  if ($logtype eq "single") {
    print CAPLOG "@capdata\n";
  }
}


lib/http.pl
# /usr/local/thp/lib/http.pl version 0.4.4
#
# httpd emultation functions for thp - Tiny Honeypot 
#
# Copyright George Bakos - alpinista@bigfoot.com
# Aud 02, 2002
# This is free software, released under the terms of the GNU General
# Public License avaiable at http://www.fsf.org/licenses/gpl.txt
#

sub http {

  while (my $commands = <STDIN>) {
    open(LOG, ">>$sesslog");
    select LOG;
    $|=1;
    print LOG $commands;
    $lcount++;
    $commands =~ s/\r//;
    my $commline = "line$lcount";
    @$commline = split /\s+/,($commands);

# Should we change labels? If selected in thp.conf, and the intruder is
# looking for common Microsoft-IIS resources, this will change the httpd 
# vendor & version to accomodate them.

    if ($line1[1] =~ /(pagerror.gif|\.asp|\.exe|\.htr|\.htx|\.htw|\.com\.dll|\.ida)[$\?%+]?/ && $chameleon eq "yes") {
      ($httpdvend, $httpdver) = ("Microsoft-IIS", "$chamelver");
    }
    $respdir = "$thpdir/lib/$httpdvend";

# Has the intruder specified an HTTP version in their request? If not,
# the session closes with an error - see err400()

    $method = $line1[0];
    $resname = $line1[1];
    $resname =~ s/^.*\///;
    $protover = "$line1[2]" if ($line1[2] =~ /HTTP\/1.[01]$/);
    
    if ($commands =~ /^$/m) {

# Check for an acceptable http method. If fatfingered or otherwise unknown,
# bomb out with an error 501. Not all daemons return 501s, some just spew
# error 400s for just about everything broken. I still need to ID where 
# this is appropriate.

        if ($method !~ /GET|POST|HEAD/ ) {
	  http_hdr("501","Bad Method","text/html");
	  err501();
	  exit 0;

# Is the URL too long? Feel free to monkey with this, or ditch it. This
# tests the entire URI, not just resource filename.

	} elsif ( length($line1[1]) > 255 ) {
	  http_hdr("414","Request-URI Too Large","text/html");
	  err414();
	  exit 0;

# Match on resource name. We allow "/" and "index.htm" and "index.html". All
# of these will return the content in lib/<vendor>/200. The return headers
# are built in http_hdr(), and content is pulled from the file. If your html
# document contains <img> tags, those image files should be placed in the same
# directory. We can't match on $resname here, since we stripped off all
# slashes, and would break default webpage requests. Thus it's back to 
# $line1[1].

	} elsif ( $line1[1] =~ m/^(\/$|\/index.htm[l]?)$/ && $protover) {
	  $respfile = "$respdir/200";
	  http_hdr("200","OK","text/html");
	  open (RESP, "$respfile");
	  while (<RESP>) {
	    chomp;
	    print STDERR ($_, "\x0d\x0a");
	  }
	  close RESP;
	  print STDERR ($_, "\x0d\x0a");
	exit 0;

# If the vendor is IIS and the request contains common default resource
# names, this returns the same lib/<vendor>/200

	} elsif ( $resname =~ /(default|iisstart|localstart)/ && $protover && $httpdvend eq "Microsoft-IIS") {
	  $respfile = "$respdir/200";
	  http_hdr("200","OK","text/html");
	  open (RESP, "$respfile");
	  while (<RESP>) {
	    chomp;
	    print STDERR ($_, "\x0d\x0a");
	  }
	  close RESP;
	  print STDERR ($_, "\x0d\x0a");
	exit 0;

# Here is the text catchall, setting a mimetype of /text/html. 

	} elsif ( -T "$respdir/$resname" && $protover) {
	  $respfile = "$respdir/$resname";
	  http_hdr("200","OK","text/html");
	  open (RESP, "$respfile");
	  while (<RESP>) {
	    print STDERR $_;
	  }
	  close RESP;
	  print STDERR "\x0d\x0a\x0d\x0a";
	exit 0;

# If the request is for an image, strip off the path and pull it out of 
# the same lib/<vendor>/ directory, modifying the mime type accordingly.

	} elsif ( $resname =~ /(gif|jpg|png)$/ && ($imgtype = "$+") && -f "$respdir/$resname" && $protover) {
	  $respfile = "$respdir/$resname";
	  http_hdr("200","OK","image/$imgtype");
	  open (RESP, "$respfile");
	  while (<RESP>) {
	    print STDERR $_;
	  }
	  close RESP;
	  print STDERR "\x0d\x0a\x0d\x0a";
	exit 0;

	} else { 
	  http_hdr("400","Bad Request","text/html");
	  err400() }
	exit 0;
    } 
    close LOG;
  }
}

sub http_hdr {
$fsize = -s $respfile;
$now = strftime("%a, %B %d %Y %T GMT", gmtime(time));
  print STDERR qq ($protover $_[0] $_[1]\x0d
Server: $httpdvend/$httpdver\x0d
Date: $now\x0d
Content-Length: $fsize\x0d
Connection: close\x0d
Content-Type: $_[2]\x0d\x0a);
if ( $_[2] =~ /image/ ) {
  print STDERR "Accept-Ranges: bytes\x0d\x0a";
}
if ( $httpdvend =~ /Microsoft/ ) {
  print STDERR "Set-Cookie: ASPSESSIONIDQQGGGHOO=GAFBCHFDEANKGFKPIPKENMAP; path=/\x0d\x0a";
  print STDERR "Cache-control: private\x0d\x0a";
}
print STDERR "\x0d\x0a";
}

sub err400 {
  my $msg = qq (<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<HTML><HEAD>
<TITLE>400 Bad Request</TITLE>
</HEAD><BODY>
<H1>Bad Request</H1>
Your browser sent a request that this server could not understand.<P>
Invalid URI in request "@line1"<P>
<HR>
<ADDRESS>$httpdvend/$httpdver Server at $thpaddr Port 80</ADDRESS>
</BODY></HTML>\x0d\x0a\x0d\x0a);
  print STDERR "$msg";
}

sub err414 {
  my $msg = qq (<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<HTML><HEAD>
<TITLE>414 Request-URI Too Large</TITLE>
</HEAD><BODY>
<H1>Request-URI Too Large</H1>
The requested URL's length exceeds the capacity
limit for this server.<P>
request failed: URI too long<P>
<HR>
<ADDRESS>$httpdvend/$httpdver Server at $thpaddr Port 80</ADDRESS>
</BODY></HTML>\x0d\x0a\x0d\x0a);
  print ( STDERR "$msg");
}

sub err501 {
  my $msg = qq (<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<HTML><HEAD>
<TITLE>501 Invalid Method</TITLE>
</HEAD><BODY>
<H1>Invalid Method</H1>
The requested method is not available on this server.<P>
request failed: Invalid or unrecognized method in "@line1"<P>
<HR>
<ADDRESS>$httpdvend/$httpdver Server at $thpaddr Port 80</ADDRESS>
</BODY></HTML>\x0d\x0a\x0d\x0a);
  print ( STDERR "$msg");
}

